Move data logic out of the inserter

noisysocks opened the pull request:
### 💃 What 

Fixes #4221. Fixes #4303.

Currently, `InserterMenu`, `VisualEditorInserter`, and `blocksAutocompleter` all directly call `getBlockTypes()` and `getReusableBlocks()` to determine what blocks should be shown.

This means that we duplicate the same logic in a few different places:

* Checking, if the block has `useOnce` set, whether it's been already used or not
* Checking whether or not the block has `isPrivate` set
* Checking whether or not the block is in the editor config's list of enabled block types
* Mixing in the list of available reusable blocks with the list of regular blocks

This approach is pretty error-prone: it's very easy to change the behaviour of one inserter and forget about another.

To start addressing this, I've changed `InserterMenu` to source its data from two new selectors:

* `getInserterItems` - used to populate the inserter's search results
* `getRecentInserterItems` - used to populate the 'Recent' tab

Both selectors filter out private/disabled blocks, correctly set `isDisabled`, and are unit tested.

### ✅ How to test

1. Insert some static blocks and reusable blocks
2. Recently inserted blocks should move to the top of the Recent tab in the main inserter

noisysocks committed:
Move data logic out of the inserter

Moves the logic that determines which items should appear in the
inserter into dedicated selector functions. This way, the logic is
easier to test and can be re-used.

youknowriad commented on editor/components/inserter/group.js:
```diff
@@ -10,71 +10,92 @@ import { Component } from '@wordpress/element';
+			this.activeItems = deriveActiveItems( nextProps.items );
 			// Try and preserve any still valid selected state.
-			const current = find( this.activeBlocks, { name: this.state.current } );
+			const current = find( this.activeItems, ( item ) => isEqual( item, this.state.current ) );
```
Can we avoid `isEqual` and use natural keys instead? (performance reasons)

noisysocks replied:
👍 fixed in ef3bbc97f96bc9c5516d7501c4039fc7a70dc47b.

youknowriad commented on editor/components/inserter/group.js:
```diff
@@ -10,71 +10,92 @@ import { Component } from '@wordpress/element';
-					block.name + block.initialAttributes.ref :
-					block.name
-				}
+				key={ index }
```
Not sure it's a good idea to use `index` here. I'd prefer a natural key as well. The block index can change when searching for instance.

noisysocks replied:
👍 fixed in ef3bbc97f96bc9c5516d7501c4039fc7a70dc47b.

youknowriad commented on editor/store/selectors.js:
```diff
@@ -1111,15 +1111,129 @@ export function getNotices( state ) {
+	);
+
+	const items = [ ...staticItems, ...dynamicItems ];
+	return compact( items );
```
Can we add a unique `id` to each item. It could equal the `name` for regular blocks and the `ref` for reusable blocks. Having a unique id across all items would simplify logic in the inserter.

noisysocks replied:
👍 fixed in ef3bbc97f96bc9c5516d7501c4039fc7a70dc47b.

youknowriad reviewed (commented):
Nice work, One concern left for me (the unique id) and this should be good to go.

noisysocks committed:
Add a unique id to inserter items

This lets the inserter do faster string comparison between inserter
items rather than deep object comparison.

youknowriad commented on editor/store/selectors.js:
```diff
@@ -1111,15 +1111,131 @@ export function getNotices( state ) {
+ * An item that appears in the inserter. Inserting this item will create a new
+ * block. Inserter items encapsulate both regular blocks and reusable blocks.
+ * 
+ * @typedef {Object} Editor.InserterItem
```
We're missing the id here ;)

noisysocks replied:
Whoops, fixed :)

youknowriad commented:
Do you plan to use these in the other insereters in a separate PR?

This looking solid for me. (Text are broken though for some reason).

noisysocks commented:
> Do you plan to use these in the other inserters in a separate PR?

Yep, I plan to use these methods in the `blocksAutocompleter`. There's no need to update `VisualEditorInserter` since it's going away.

aduth commented on editor/store/selectors.js:
```diff
@@ -1111,15 +1111,132 @@ export function getNotices( state ) {
+		return null;
+	}
+
+	const blockTypeIsDisabled = Array.isArray( enabledBlockTypes ) && ! enabledBlockTypes.includes( blockType.name );
```
ES6 prototype methods are not polyfilled (e.g. `Array.from` is fine, but `Array#includes` is not).

This will error in IE11.

noisysocks replied:
👌 good to know. Looks like https://github.com/WordPress/gutenberg/pull/5528 addresses this.

noisysocks replied:
Could we polyfill prototype methods?

aduth replied:
Context at #746

tl;dr: Was waiting for Babel 7, which I didn't expect to take so long to release 😄
